File: app/src/main/java/bassamalim/hidaya/activities/RadioClient.java
package bassamalim.hidaya.activities;

import android.content.ComponentName;
import android.content.SharedPreferences;
import android.media.AudioManager;
import android.os.Bundle;
import android.support.v4.media.MediaBrowserCompat;
import android.support.v4.media.MediaMetadataCompat;
import android.support.v4.media.session.MediaControllerCompat;
import android.support.v4.media.session.MediaSessionCompat;
import android.support.v4.media.session.PlaybackStateCompat;
import android.util.Log;
import android.widget.ImageButton;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.res.ResourcesCompat;
import androidx.preference.PreferenceManager;

import com.google.firebase.remoteconfig.FirebaseRemoteConfig;

import bassamalim.hidaya.R;
import bassamalim.hidaya.databinding.ActivityRadioClientBinding;
import bassamalim.hidaya.other.Global;
import bassamalim.hidaya.other.Utils;
import bassamalim.hidaya.services.RadioService;

public class RadioClient extends AppCompatActivity {

    private ActivityRadioClientBinding binding;
    private SharedPreferences pref;
    private FirebaseRemoteConfig remoteConfig;
    private MediaBrowserCompat mediaBrowser;
    private MediaControllerCompat controller;
    private MediaControllerCompat.TransportControls tc;
    private ImageButton ppBtn;    // play/pause button
    private String theme;
    private String link;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Utils.onActivityCreateSetTheme(this);
        binding = ActivityRadioClientBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.home.setOnClickListener(v -> onBackPressed());

        pref = PreferenceManager.getDefaultSharedPreferences(this);

        initViews();

        mediaBrowser = new MediaBrowserCompat(this, new ComponentName(this,
                RadioService.class), connectionCallbacks, null); // optional Bundle

        getLinkAndEnable();
    }

    @Override
    public void onStart() {
        super.onStart();
        mediaBrowser.connect();
    }

    @Override
    public void onResume() {
        super.onResume();
        setVolumeControlStream(AudioManager.STREAM_MUSIC);
    }

    @Override
    public void onStop() {
        super.onStop();
        if (MediaControllerCompat.getMediaController(RadioClient.this) != null) {
            MediaControllerCompat.getMediaController(RadioClient.this)
                    .unregisterCallback(controllerCallback);
        }
        mediaBrowser.disconnect();
    }

    private void initViews() {
        theme = pref.getString(getString(R.string.theme_key), "ThemeM");

        ppBtn = binding.radioPpBtn;
    }

    private void getLinkAndEnable() {
        remoteConfig = FirebaseRemoteConfig.getInstance();
        remoteConfig.fetchAndActivate().addOnCompleteListener(this, task -> {
            if (task.isSuccessful()) {
                link = remoteConfig.getString("quran_radio_url");

                Log.d(Global.TAG, "Config params updated");
                Log.d(Global.TAG, "Quran Radio URL: " + link);

                enableControls();
            }
            else
                Log.d(Global.TAG, "Fetch failed");
        });
    }

    private final MediaBrowserCompat.ConnectionCallback connectionCallbacks =
            new MediaBrowserCompat.ConnectionCallback() {
        @Override
        public void onConnected() {
            // Get the token for the MediaSession
            MediaSessionCompat.Token token = mediaBrowser.getSessionToken();

            // Create a MediaControllerCompat
            MediaControllerCompat mediaController = new MediaControllerCompat(
                    RadioClient.this, token);

            // Save the controller
            MediaControllerCompat.setMediaController(RadioClient.this, mediaController);
            controller = MediaControllerCompat.getMediaController(RadioClient.this);
            tc = controller.getTransportControls();

            // Finish building the UI
            buildTransportControls();
        }

        @Override
        public void onConnectionSuspended() {
            Log.e(Global.TAG, "Connection suspended in RadioClient");
            // The Service has crashed.
            // Disable transport controls until it automatically reconnects
            disableControls();
        }

        @Override
        public void onConnectionFailed() {
            Log.e(Global.TAG, "Connection failed in RadioClient");
            // The Service has refused our connection
            disableControls();
        }
    };

    MediaControllerCompat.Callback controllerCallback = new MediaControllerCompat.Callback() {
        @Override
        public void onMetadataChanged(MediaMetadataCompat metadata) {
            // To change the metadata inside the app when the user changes it from the notification
            //updateMetadata(metadata);
        }

        @Override
        public void onPlaybackStateChanged(PlaybackStateCompat state) {
            // To change the playback state inside the app when the user changes it
            // from the notification
            updatePbState(state);
        }
    };

    private void buildTransportControls() {
        //enableControls();

        // Display the initial state
        updatePbState(controller.getPlaybackState());

        // Register a Callback to stay in sync
        controller.registerCallback(controllerCallback);
    }

    private void updatePbState(PlaybackStateCompat state) {
        int currentState = state.getState();
        updateButton(currentState == PlaybackStateCompat.STATE_PLAYING);
    }

    private void updateButton(boolean playing) {
        if (theme.equals("ThemeM")) {
            if (playing)
                ppBtn.setImageDrawable(ResourcesCompat.getDrawable(getResources(),
                        R.drawable.ic_player_pause, getTheme()));
            else
                ppBtn.setImageDrawable(ResourcesCompat.getDrawable(getResources(),
                        R.drawable.ic_player_play, getTheme()));
        }
        else {
            if (playing)
                ppBtn.setImageDrawable(ResourcesCompat.getDrawable(getResources(),
                        R.drawable.ic_player_pause_l, getTheme()));
            else
                ppBtn.setImageDrawable(ResourcesCompat.getDrawable(getResources(),
                        R.drawable.ic_player_play_l, getTheme()));
        }
    }

    private void enableControls() {
        // Attach a listeners to the buttons
        ppBtn.setOnClickListener(v -> {
            // Since this is a play/pause button, you'll need to test the current state
            // and choose the action accordingly
            int pbState = controller.getPlaybackState().getState();

            if (pbState == PlaybackStateCompat.STATE_PLAYING)
                tc.pause();
            else
                tc.playFromMediaId(link, null);
        });
    }

    private void disableControls() {
        ppBtn.setOnClickListener(null);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        binding = null;
        mediaBrowser = null;
    }
}

File: app/src/main/java/bassamalim/hidaya/activities/RadioClient.kt
package bassamalim.hidaya.activities

import android.content.ComponentName
import android.media.AudioManager
import android.os.Build
import android.os.Bundle
import android.support.v4.media.MediaBrowserCompat
import android.support.v4.media.MediaMetadataCompat
import android.support.v4.media.session.MediaControllerCompat
import android.support.v4.media.session.PlaybackStateCompat
import android.util.Log
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.annotation.RequiresApi
import androidx.compose.foundation.layout.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import bassamalim.hidaya.R
import bassamalim.hidaya.other.Global
import bassamalim.hidaya.services.RadioService
import bassamalim.hidaya.ui.components.MyPlayerBtn
import bassamalim.hidaya.ui.components.MyScaffold
import bassamalim.hidaya.ui.components.MyText
import bassamalim.hidaya.ui.theme.AppTheme
import bassamalim.hidaya.utils.ActivityUtils
import com.google.firebase.remoteconfig.FirebaseRemoteConfig

@RequiresApi(Build.VERSION_CODES.O)
class RadioClient : ComponentActivity() {

    private lateinit var remoteConfig: FirebaseRemoteConfig
    private var mediaBrowser: MediaBrowserCompat? = null
    private lateinit var controller: MediaControllerCompat
    private lateinit var tc: MediaControllerCompat.TransportControls
    private lateinit var url: String
    private val btnState = mutableStateOf(PlaybackStateCompat.STATE_NONE)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        ActivityUtils.myOnActivityCreated(this)

        setContent {
            AppTheme {
                UI()
            }
        }

        getLinkAndConnect()
    }

    override fun onStart() {
        super.onStart()
        mediaBrowser?.connect()
    }

    override fun onResume() {
        super.onResume()
        volumeControlStream = AudioManager.STREAM_MUSIC
    }

    override fun onStop() {
        super.onStop()
        if (MediaControllerCompat.getMediaController(this@RadioClient) != null) {
            MediaControllerCompat.getMediaController(this@RadioClient)
                .unregisterCallback(controllerCallback)
        }
        mediaBrowser?.disconnect()
    }

    private fun getLinkAndConnect() {
        remoteConfig = FirebaseRemoteConfig.getInstance()
        remoteConfig.fetchAndActivate().addOnCompleteListener(this) { task ->
            if (task.isSuccessful) {
                url = remoteConfig.getString("quran_radio_url")
                Log.i(Global.TAG, "Config params updated")
                Log.i(Global.TAG, "Quran Radio URL: $url")

                connect()
            }
            else Log.e(Global.TAG, "Fetch failed")
        }
    }

    private fun connect() {
        mediaBrowser = MediaBrowserCompat(
            this, ComponentName(this, RadioService::class.java),
            connectionCallbacks, null
        )
        mediaBrowser!!.connect()
    }

    private val connectionCallbacks: MediaBrowserCompat.ConnectionCallback =
        object : MediaBrowserCompat.ConnectionCallback() {
        override fun onConnected() {
            // Get the token for the MediaSession
            val token = mediaBrowser!!.sessionToken

            // Create a MediaControllerCompat
            val mediaController = MediaControllerCompat(this@RadioClient, token)

            // Save the controller
            MediaControllerCompat.setMediaController(this@RadioClient, mediaController)
            controller = MediaControllerCompat.getMediaController(this@RadioClient)
            tc = controller.transportControls

            // just sending the static url to extract the dynamic url
            tc.playFromMediaId(url, null)

            // Finish building the UI
            buildTransportControls()

            btnState.value = PlaybackStateCompat.STATE_STOPPED
        }

        override fun onConnectionSuspended() {
            Log.e(Global.TAG, "Connection suspended in RadioClient")
            // The Service has crashed.
            // Disable transport controls until it automatically reconnects
            btnState.value = PlaybackStateCompat.STATE_NONE
        }

        override fun onConnectionFailed() {
            Log.e(Global.TAG, "Connection failed in RadioClient")
            // The Service has refused our connection
            btnState.value = PlaybackStateCompat.STATE_NONE
        }
    }

    private var controllerCallback: MediaControllerCompat.Callback =
        object : MediaControllerCompat.Callback() {
        override fun onMetadataChanged(metadata: MediaMetadataCompat?) {
            // To change the metadata inside the app when the user changes it from the notification
            //updateMetadata(metadata);
        }

        override fun onPlaybackStateChanged(state: PlaybackStateCompat) {
            // To change the playback state inside the app when the user changes it
            // from the notification
            updatePbState(state)
        }
    }

    private fun buildTransportControls() {
        // Display the initial state
        updatePbState(controller.playbackState)

        // Register a Callback to stay in sync
        controller.registerCallback(controllerCallback)
    }

    private fun updatePbState(state: PlaybackStateCompat) {
        when (state.state) {
            PlaybackStateCompat.STATE_PLAYING,
            PlaybackStateCompat.STATE_STOPPED,
            PlaybackStateCompat.STATE_CONNECTING -> btnState.value = state.state
            else -> {}
        }
    }

    @Composable
    private fun UI() {
        MyScaffold(stringResource(id = R.string.quran_radio)) {
            Column(
                Modifier
                    .fillMaxSize()
                    .padding(it),
                verticalArrangement = Arrangement.Center,
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                MyText(
                    text = stringResource(R.string.holy_quran_radio),
                    fontSize = 30.sp,
                    fontWeight = FontWeight.Bold,
                    modifier = Modifier.padding(bottom = 50.dp)
                )

                MyPlayerBtn(state = btnState, padding = 10.dp) {
                    if (btnState.value != PlaybackStateCompat.STATE_NONE) {
                        // Since this is a play/pause button
                        // test the current state and choose the action accordingly=
                        if (controller.playbackState.state == PlaybackStateCompat.STATE_PLAYING) {
                            tc.pause()
                            btnState.value = PlaybackStateCompat.STATE_STOPPED
                        }
                        else {
                            tc.play()
                            btnState.value = PlaybackStateCompat.STATE_PLAYING
                        }
                    }
                }
            }
        }
    }
}